Perform a Chi-square goodness-of-fit test for a categorical variable. Interpret whether the
observed distribution aligns with an expected distribution.

In [126]:
import numpy as np
import pandas as pd

In [127]:
df = pd.read_csv(r'E:\Code\Maths assignment\tested.csv')

In [128]:
newdf = df[['Survived', 'Embarked']]

In [ ]:
newdf = newdf.groupby(['Embarked', 'Survived']).size().unstack().reset_index()
newdf.columns.name = None

Example: If historical records suggest 60% embarked from S, 25% from C, and 15% from Q, the expected count would be:

In [132]:
from scipy.stats import chisquare

# Observed frequencies
observed = newdf[[0, 1]].sum(axis=1).values

# Expected frequencies based on historical proportions
total = observed.sum()
expected = total * np.array([0.25, 0.15, 0.60])  # Proportions for C, Q, S

# Perform the Chi-square goodness-of-fit test
chi2_stat, p_value = chisquare(f_obs=observed, f_exp=expected)

print(f"Chi-square Statistic: {chi2_stat}")
print(f"P-value: {p_value}")

Chi-square Statistic: 5.977671451355663
P-value: 0.05034601919131072


The p-value for the Chi-square goodness-of-fit test is 0.0503, which is slightly above the commonly used significance level of 0.05. This means we fail to reject the null hypothesis. There is not enough evidence to conclude that the observed distribution of passengers across Embarked locations (S, C, Q) significantly differs from the historical distribution (60%, 25%, 15%). Thus, the observed pattern is reasonably consistent with the expected one.